plugins/devlink/fu-devlink-plugin.c

```diff
@@ -0,0 +1,341 @@
+	physical_id = g_strdup_printf("%s/%s", bus_name, dev_name);
+	fu_plugin_cache_add(FU_PLUGIN(self), physical_id, device);
+	fu_plugin_device_add(FU_PLUGIN(self), device);
+}
+
+static void
+fu_devlink_plugin_device_removed_from_netlink(FuDevlinkPlugin *self, const struct nlmsghdr *nlh)
+{
+	struct nlattr *tb[DEVLINK_ATTR_MAX + 1] = {};
+	g_autofree gchar *physical_id = NULL;
+	const gchar *bus_name = NULL;
+	const gchar *dev_name = NULL;
+	FuDevice *device = NULL;
+
+	/* Parse netlink attributes using libmnl */
+	mnl_attr_parse(nlh, sizeof(struct genlmsghdr), fu_devlink_netlink_attr_cb, tb);
+
+	/* Extract bus and device names */
+	if (tb[DEVLINK_ATTR_BUS_NAME])
+		bus_name = mnl_attr_get_str(tb[DEVLINK_ATTR_BUS_NAME]);
+	if (tb[DEVLINK_ATTR_DEV_NAME])
+		dev_name = mnl_attr_get_str(tb[DEVLINK_ATTR_DEV_NAME]);
+
+	if (!bus_name || !dev_name) {
+		g_debug("devlink device removal notification missing bus_name or dev_name");
+		return;
+	}
+
+	g_debug("devlink device removed: %s/%s", bus_name, dev_name);
+
+	/* Find and remove device */
+	physical_id = g_strdup_printf("%s/%s", bus_name, dev_name);
+	device = fu_plugin_cache_lookup(FU_PLUGIN(self), physical_id);
+	if (device) {
+		fu_plugin_cache_remove(FU_PLUGIN(self), physical_id);
+		fu_plugin_device_remove(FU_PLUGIN(self), device);
+	}
+}
+
+/* Callback for processing individual netlink messages */
+static int
+fu_devlink_plugin_process_message_cb(const struct nlmsghdr *nlh, void *data)
+{
+	FuDevlinkPlugin *self = FU_DEVLINK_PLUGIN(data);
+	struct genlmsghdr *genl;
+
+	if (nlh->nlmsg_type == NLMSG_ERROR) {
+		g_debug("devlink netlink error message");
+		return MNL_CB_OK;
+	}
+
+	if (nlh->nlmsg_type != self->nlg.family_id)
+		return MNL_CB_OK;
+
+	genl = mnl_nlmsg_get_payload(nlh);
+	switch (genl->cmd) {
+	case DEVLINK_CMD_NEW:
+		fu_devlink_plugin_device_added_from_netlink(self, nlh);
+		break;
+	case DEVLINK_CMD_DEL:
+		fu_devlink_plugin_device_removed_from_netlink(self, nlh);
+		break;
+	default:
+		break;
+	}
+
+	return MNL_CB_OK;
+}
+
+static gboolean
+fu_devlink_plugin_netlink_cb(GIOChannel *channel, GIOCondition condition, gpointer user_data)
+{
+	FuDevlinkPlugin *self = FU_DEVLINK_PLUGIN(user_data);
+	const size_t buf_size = 8192;
+	g_autofree gchar *buf = g_malloc0(buf_size);
+	gsize len;
+	GIOStatus status;
+	GError *error = NULL;
+	gint rc;
+
+	if (condition & (G_IO_ERR | G_IO_HUP)) {
+		g_debug("devlink netlink socket error");
+		return FALSE;
+	}
+
+	/* Read netlink message via GIOChannel */
+	status = g_io_channel_read_chars(channel, buf, buf_size, &len, &error);
+	if (status != G_IO_STATUS_NORMAL) {
+		if (error) {
+			g_debug("failed to read devlink netlink message: %s", error->message);
+			g_error_free(error);
+		}
+		return TRUE;
+	}
+
+	/* Process netlink messages using libmnl */
+	rc = mnl_cb_run(buf, len, 0, 0, fu_devlink_plugin_process_message_cb, self);
+	if (rc == -1)
+		g_debug("failed to process netlink message: %s", fwupd_strerror(errno));
```

Comment:
g_warning perhaps?

Reply:
I forgot to rewrite this direct mnl_cb_run call to netlink wrappers. Did that, error is now filled-up so I'm using "g_warning("failed to process netlink message: %s", error->message);"